pkg/admin/courseserver/courseserver.go

```diff
@@ -0,0 +1,393 @@
+	s := AdminCourseServer{}
+
+	s.hfClientSet = hfClientset
+	s.auth = authClient
+
+	return &s, nil
+}
+
+func (a AdminCourseServer) SetupRoutes(r *mux.Router) {
+	r.HandleFunc("/a/course/list", a.ListFunc).Methods("GET")
+	r.HandleFunc("/a/course/{id}", a.GetFunc).Methods("GET")
+	r.HandleFunc("/a/course/new", a.CreateFunc).Methods("POST")
+	r.HandleFunc("/a/course/{id}", a.UpdateFunc).Methods("PUT")
+	r.HandleFunc("/a/course/{id}", a.DeleteFunc).Methods("DELETE")
+}
+
+func (a AdminCourseServer) GetFunc(w http.ResponseWriter, r *http.Request) {
+	_, err := a.auth.AuthNAdmin(w, r)
+	if err != nil {
+		util.ReturnHTTPMessage(w, r, 403, "forbidden", "no access to get course")
+		return
+	}
+
+	vars := mux.Vars(r)
+
+	id := vars["id"]
+
+	if id == "" {
+		util.ReturnHTTPMessage(w, r, 400, "badrequest", "no id passed in")
+		return
+	}
+
+	obj, err := a.hfClientSet.HobbyfarmV1().Courses().Get(id, metav1.GetOptions{})
+	if err != nil {
+		util.ReturnHTTPMessage(w, r, 400,"badrequest", "invalid id")
+		return
+	}
+
+	preparedCourse := PreparedCourse{obj.Name, obj.Spec}
+
+	encodedCourse, err := json.Marshal(preparedCourse)
+	if err != nil {
+		glog.Errorf("error encoding course: %v", err)
+		util.ReturnHTTPMessage(w, r, 500, "internalerror", "error retrieving course")
+		return
+	}
+
+	util.ReturnHTTPContent(w, r, 200, "success", encodedCourse)
+
+	glog.V(2).Infof("retrieve course %s", obj.Name)
+}
+
+func (a AdminCourseServer) ListFunc(w http.ResponseWriter, r *http.Request) {
+	_, err := a.auth.AuthNAdmin(w, r)
+	if err != nil {
+		util.ReturnHTTPMessage(w, r, 403, "forbidden", "no access to list courses")
+		return
+	}
+
+	tempCourses, err := a.hfClientSet.HobbyfarmV1().Courses().List(metav1.ListOptions{})
+	if err != nil {
+		glog.Errorf("error listing courses: %v", err)
+		util.ReturnHTTPMessage(w, r, 500, "internalerror", "error listing courses")
+		return
+	}
+
+	var courses []PreparedCourse
+	for _, c := range tempCourses.Items {
+		courses = append(courses, PreparedCourse{ c.Name, c.Spec})
+	}
+
+	encodedCourses, err := json.Marshal(courses)
+	if err != nil {
+		glog.Errorf("error marshalling prepared courses: %v", err)
+		util.ReturnHTTPMessage(w, r, 500, "internalerror", "error listing courses")
+		return
+	}
+
+	util.ReturnHTTPContent(w, r, 200, "success", encodedCourses)
+
+	glog.V(2).Infof("listed courses")
+
+}
+
+func (a AdminCourseServer) CreateFunc(w http.ResponseWriter, r *http.Request) {
+	_, err := a.auth.AuthNAdmin(w, r)
+	if err != nil {
+		util.ReturnHTTPMessage(w, r, 403, "forbidden", "no access to create courses")
+		return
+	}
+
+	name := r.PostFormValue("name")
+	if name == "" {
+		util.ReturnHTTPMessage(w, r, 400, "badrequest", "no name passed in")
+		return
+	}
+
+	description := r.PostFormValue("description")
+	if description == "" {
+		util.ReturnHTTPMessage(w, r, 400, "badrequest", "no description passed in")
+		return
+	}
+
+	keepaliveDuration := r.PostFormValue("keepalive_duration")
```

Comment:
just as a side thought, I wonder if we should validate these values against malicious users.

Reply:
i.e. regex them?

Reply:
regex would be one way to do it, and probably safe enough™

Reply:
do you want that part of this commit, or handled as part of a project-wide strategy to regex

Reply:
e.g. new PR specifically for that

Reply:
@ebauman that's up to you. probably a separate commit makes sense, I was just making a comment